canal/sync.go

```diff
@@ -336,3 +341,37 @@ func (c *Canal) CatchMasterPos(timeout time.Duration) error {
 
 	return c.WaitUntilPos(pos, timeout)
 }
+
+type Position struct {
```

Comment:
why do we need it?

Reply:
Based on `mysql.Position` adds `SavePos` and `Force` fields

I think queryEvent contains multiple sub-events, pos save logic, and it is up to the end-user event to decide whether to save or not

---
<img width="241" alt="image" src="https://github.com/go-mysql-org/go-mysql/assets/8416556/4cd530a2-4b6f-4114-a141-8ad77b3297ad">


``` go

func (h *DummyEventHandler) OnQueryEvent(header *replication.EventHeader, stmt ast.StmtNode, pos *Position, e *replication.QueryEvent) error { 
         //....
        pos.SavePos =true
	return nil
}

```

---
Of course, it can also be handled uniformly in the main loop of canal/sync.go events.


``` go
func (c *Canal) startSyncer() (*replication.BinlogStreamer, error) { 
        //....
       for _, stmt := range stmts {
		nodes := parseDDLStmt(stmt)
		if len(nodes) > 0 {
		      // OnDDL
		} else {
			err := c.handleQueryEvent(ev.Header, stmt, posInfo, e)
			if err != nil {
				c.cfg.Logger.Errorf("handle query event err %v", err)
			} else {
				savePos = true   // for save pos
				force = true   // for save pos
			}
               }
              //....
       } 
      //....
}
```

Reply:
> it is up to the end-user event to decide whether to save or not

I think it's more obvious to use return value to do it.

```
OnQueryEvent(ev *replication.BinlogEvent, e *replication.QueryEvent, stmt ast.StmtNode, pos mysql.Position) (savePos bool, force bool, err error)
```

Reply:
Ok, I'll modify it